### ISSUE TITLE
Incorrect gradient with MX conditional with constant expression

### ISSUE THREAD

**Issue by BachiLi** [casadi/casadi#1553]
If "conditional" contains a constant expression in one of the branches, casadi seems to treat the whole conditional as a constant and fail to compute correct derivative.

For example, following code outputs incorrect gradient function:

```
from casadi import *
c = MX.sym('c')
x = MX.sym('x')
z = conditional(c, [sin(x), cos(x)], 0)

test_switch = MXFunction("test_switch", [x, c],[z])
test_switch_grad = test_switch.gradient(0, 0)

print test_switch_grad
```

The output is:

```
 Number of inputs: 2
  Input 0, a.k.a. "i0", 1-by-1 (dense), No description available
  Input 1, a.k.a. "i1", 1-by-1 (dense), No description available
 Number of outputs: 2
  Output 0, a.k.a. "grad", 1-by-1 (dense), No description available
  Output 1, a.k.a. "o0", 1-by-1 (dense), No description available
@0 = 0
output[0] = @0
@0 = input[1][0]
@1 = input[0][0]
@2 = conditional(@0, @1)
output[1] = @2
```

The function always output 0 for the gradient, which is obviously incorrect (I am using the develop branch).

**Comment by jaeandersson** [comment]
I think it's the expected behaviour. You have the function "f(c, x, y) = c ? x : y". The derivative with respect to c is is 0 almost everywhere (for piecewise functions, CasADi calculates derivatives in the "almost everywhere" sense).

**Comment by BachiLi** [comment]
I disagree.  I was taking the derivative with respect to "x" instead of "c" here.
```
test_switch = MXFunction("test_switch", [x, c],[z])
test_switch_grad = test_switch.gradient(0, 0)
```

Notice that x is the 0-th parameter.

**Comment by jaeandersson** [comment]
Right you are. This looks like a legitimate bug.

**Comment by jgillis** [comment]
Fixed in 3.1
